athena-elasticsearch/src/main/java/com/amazonaws/athena/connectors/elasticsearch/ElasticsearchRecordHandler.java

```diff
@@ -147,38 +151,54 @@ protected void readWithConstraint(BlockSpiller spiller, ReadRecordsRequest recor
-
-                    // Throw on query timeout.
+                SearchRequest searchRequest = new SearchRequest(index)
+                        .preference(shard)
+                        .scroll(TimeValue.timeValueSeconds(60))
+                        .source(searchSource.from(0));
+
+                //init scroll
+                Scroll scroll = new Scroll(TimeValue.timeValueSeconds(60));
+
+                //Read the returned scroll id, which points to the search context that’s being kept alive and will be needed in the following search scroll call
+                SearchResponse searchResponse = client.search(searchRequest, RequestOptions.DEFAULT);
+                String scrollId = searchResponse.getScrollId();
+                SearchHit[] searchHits = searchResponse.getHits().getHits();
+                Iterator<SearchHit> iterator = searchResponse.getHits().iterator();
+
+                while (searchHits != null && searchHits.length > 0 && queryStatusChecker.isQueryRunning()) {
+                    Iterator<SearchHit> finalIterator = iterator;
+                    while (finalIterator.hasNext() && queryStatusChecker.isQueryRunning()) {
+                        ++numRows;
+                        spiller.writeRows((Block block, int rowNum) ->
+                                rowWriter.writeRow(block, rowNum, client.getDocument(finalIterator.next())) ? 1 : 0);
+                    }
+
+                    //prep for next hits and keep track of scroll id.
+                    SearchScrollRequest scrollRequest = new SearchScrollRequest(scrollId);
+                    scrollRequest.scroll(scroll);
+                    searchResponse = client.scroll(scrollRequest, RequestOptions.DEFAULT);
                     if (searchResponse.isTimedOut()) {
                         throw new RuntimeException("Request for index (" + index + ") " + shard + " timed out.");
                     }
 
-                    // Increment current position to next batch of results.
-                    currPosition += QUERY_BATCH_SIZE;
-                    // Process hits.
-                    Iterator<SearchHit> hitIterator = searchResponse.getHits().iterator();
-                    hitsNum = searchResponse.getHits().getHits().length;
-
-                    while (hitIterator.hasNext() && queryStatusChecker.isQueryRunning()) {
-                        ++numRows;
-                        spiller.writeRows((Block block, int rowNum) ->
-                                rowWriter.writeRow(block, rowNum, client.getDocument(hitIterator.next())) ? 1 : 0);
+                    if (searchResponse.getHits() == null) {
+                        break;
                     }
-                    // if hitsNum < QUERY_BATCH_SIZE, then this is the last batch of documents.
-                } while (hitsNum == QUERY_BATCH_SIZE && queryStatusChecker.isQueryRunning());
+                    scrollId = searchResponse.getScrollId();
```

Comment:
wondering if we should put this at the top of the loop and remove the assignments that occur before the loop. That way we don't repeat ourselves and risk a bug introduced from the copy. The only thing is that you then need to break up the while condition into a break clause.

style wise, deferring to your judgment, just raising the question